### DISCUSSION TITLE
Simple questions can be asked here

### DISCUSSION THREAD

**Discussion by 510ren** [510ren/ytdl-core#13] [General]
If you have any quick questions, send them here. We will reply about it.

> [!NOTE]
> Before submitting here, please check the FAQs on the Wiki for questions. 
> If the question is deemed necessary to create another issue, we will reply "Please create a new issue".

**Comment by awnigharbia** [comment]


**Reply by 510ren** [reply]
Try removing the clients specification; all video URLs retrieved from the web client will return 403.

**Reply by awnigharbia** [reply]
I got my server ip banned. So it requires me to sign in.

[ DEBUG ]: [ web ]: Success
[info] [ DEBUG ]: [ webCreator ]: Success
 [info] [ DEBUG ]: [ tvEmbedded ]: Error
 [info] Reason: Sign in to confirm you’re not a bot
 [info] [ DEBUG ]: [ ios ]: Error
 [info] Reason: Sign in to confirm you’re not a bot
 [info] [ DEBUG ]: [ android ]: Error
 [info] Reason: Sign in to confirm you’re not a bot

How to resolve the sign in issue?

**Reply by 510ren** [reply]
It may be useful to adapt OAuth2 with reference to README.

**Reply by awnigharbia** [reply]
Okay I'll use it and give you feedback ❤️.

I hope you find a way to parse `m3u8_native`. it'll be so usefull with this awesome package.

**Reply by awnigharbia** [reply]
I did used oath2 and now they are all passed. but I still get 403 error for this video id `gQXha8JzcGs`

```
const NORMARL_OATH2 = new YtdlCore.OAuth2({...  });

            const ytdl = new YtdlCore({
                oauth2: NORMARL_OATH2,
            });


            const download = ytdl.download(videoId, {
                oauth2: NORMARL_OATH2,
                quality: 'highestvideo',
                filter: (format) => {
                    try {
                        const qualityLabel = format.quality.label;
                        const indexOfP = qualityLabel.indexOf('p');
                        const qualityNumber = parseInt(qualityLabel.slice(0, indexOfP));
                        const containsHDR = qualityLabel.toLowerCase().includes('hdr');
                        const higherThan1080 = qualityNumber > 1080;

                        if (containsHDR || higherThan1080) {
                            return false;
                        }

                        return true;
                    } catch (e) {
                        return false;
                    }
                }
            })
```

How to solve it?

**Reply by 510ren** [reply]
We have just released v5.1.1, so please update.

**Reply by awnigharbia** [reply]
I did update it and still the same error.

**Reply by 510ren** [reply]
Please tell me what kind of logs are output by specifying YTDL_DEBUG in .env (environment variable).

**Reply by awnigharbia** [reply]
Note: it works well for other videos but with this one it have issues.

**Reply by 510ren** [reply]
That video also works in my environment. (You can get the URL below)

**Reply by awnigharbia** [reply]
Well as you may noticed I am selecting 1080p formats. and the 360p working well

**Reply by awnigharbia** [reply]
[ WARNING ]: Specify poToken for stable and fast operation. See README for details.
 [info] [ INFO! ]: Automatically generates poToken, but stable operation cannot be guaranteed.
 [info] [ SUCCESS ]: Successfully generated a poToken.
[info] [ INFO! ]: Since PoToken and VisitorData are not specified, they are generated automatically.
[info] [ WARNING ]: Specify poToken for stable and fast operation. See README for details.
 [info] [ INFO! ]: Automatically generates poToken, but stable operation cannot be guaranteed.
 [info] [ SUCCESS ]: Successfully generated a poToken.
 [info] [ SUCCESS ]: Successfully generated a poToken.
 [info] [ DEBUG ]: [ web ]: Success
 [info] [ DEBUG ]: [ webCreator ]: Success
 [info] [ DEBUG ]: [ tvEmbedded ]: Success
 [info] [ DEBUG ]: [ ios ]: Success
 [info] [ DEBUG ]: [ android ]: Success
 [info] [ DEBUG ]: [ web ]: Success
 [info] Download audio completed.
[info] Error occurred during the download: Request failed with status code 403

**Reply by 510ren** [reply]
Could you try specifying `filteringClients: ["webCreator"]` as an argument to the downlaod function?
```ts
ytdl.download(VIDEO_URL, {
    filteringClients: ["webCreator"],
});
```

**Reply by awnigharbia** [reply]
I did and set `disableDefaultClients: true` and it doesn't work also.

**Reply by 510ren** [reply]
> I did and set `disableDefaultClients: true` and it doesn't work also.

How are audio and video downloads performed? I could only find the DOWNLOAD function in the code.

**Reply by awnigharbia** [reply]
I'll filter out not `webCreator` from formatFilter function. and give it a shot.

**Reply by 510ren** [reply]
> I'll filter out not `webCreator` from formatFilter function. and give it a shot.

Can you try to exclude web clients?
403s are often generated by web clients.

**Reply by awnigharbia** [reply]
the audio have no problems. but the video have issues.

**Reply by awnigharbia** [reply]
You mean web and webcreators?

> > I'll filter out not `webCreator` from formatFilter function. and give it a shot.
> 
> Can you try to exclude web clients? 403s are often generated by web clients.

**Reply by 510ren** [reply]
> You mean web and webcreators?

There is no need to exclude webCreator.

**Reply by awnigharbia** [reply]
How does the package choose from links ?

**Reply by awnigharbia** [reply]
I think I'll filter out links that doesn't have `200 OK` status code.

**Reply by 510ren** [reply]
If possible, could you send the entire implementation code?
I am concerned that the DOWNLOAD function is only executed once.

**Reply by awnigharbia** [reply]
I did solve it. but it was so complex. I was using recursive to loop through 403 links until I download the video or audio. Thx bud.

**Reply by 510ren** [reply]
Good. This conversation has been resolved and will be hidden.

**Comment by MOXHARTZ** [comment]
Hi, I am getting a 403 error, but the responses on all platforms are showing as `Success`. When I change my ip this problem solves, but  after converting `35868` songs, I start getting 403 error again. I used oAuth but it still gave the same error. Can you help me with this issue?

![image](https://github.com/user-attachments/assets/c492efe1-e123-4c57-a729-7d0066f4738b)

```
const youtubeConverter = async (url, res) => {
    try {
        const soundId = new Url().getVideoID(url)
        const data = await ytdlProxy(url)
        if (data === 'local') return res.send({
            status: 'done'
        })
        if (!data) {
            return res.send({
                error: 'Video is not found',
                errorCode: 6,
                data: null
            })
        }
        if (data === 'duration') {
            return res.send({
                error: 'Video is longer than 15 minutes',
                errorCode: 5,
                data: null
            })
        }

        const writeStream = fs.createWriteStream(path.join(process.cwd(), '/audios/', soundId + '.mp3'))
        data.pipe(writeStream)
        console.log('youtubeConverter ::: Writing stream to file', url)
        data.on('end', () => res.send({ status: 'done' }))
        data.on('error', (e) => {
            console.error('youtubeConverter ::: error URL', url, 'Error:', e)
            checkError(url, e)
            deleteCache(url)
            deleteCacheFromJSON()
            fs.stat(path.join(process.cwd(), '/audios/', soundId + '.mp3'), (err, stats) => {
                if (err) {
                    console.error('fs.stat error', err)
                    return
                }
                fs.unlink(path.join(process.cwd(), '/audios/', soundId + '.mp3'), () => {
                    fs.close(writeStream.fd)
                    console.error('Youtube Audio Deleted', soundId, 'because of failed download')
                })
            })

            res.send({
                error: '' + e,
                errorCode: 7,
                data: null
            })
        })
    } catch (e) {
        console.error('youtubeConverter ::: error URL', url, 'Error:', e)
        checkError(url, e)
        res.send({
            error: '' + e,
            errorCode: 7,
            data: null
        })
    }
};

const ytdlProxy = async (url, customUrl, customProxy) => {
    const soundId = new Url().getVideoID(url)
    if (!soundId) return false;
    const uri = customUrl ?? url
    const hasLocal = fs.existsSync(path.join(process.cwd(), '/audios/', soundId + '.mp3'))
    if (hasLocal) {
        const cached = getSoundFromCache(url)
        if (!cached) {
            console.log('ytdlProxy ::: local file is served but not cached', url, 'It will be cached now')
            saveCache(uri, {
                videoId: soundId,
                created: new Date().getTime(),
            })
        }
        const stat = fs.statSync(path.join(process.cwd(), '/audios/', soundId + '.mp3'))
        if (stat.size === 0) {
            fs.unlinkSync(path.join(process.cwd(), '/audios/', soundId + '.mp3'))
            deleteCache(url)
            deleteCacheFromJSON()
            console.error('ytdlProxy Deleted', soundId, 'because of empty file')
            return false;
        }
        console.log('ytdlProxy ::: local file is served', url)
        return 'local'
    }
    const poToken = GetPOToken()
    const visitorData = GetVisitorData()
    const ytdl = new YtdlCore({
        poToken: poToken,
        visitorData: visitorData,
    })
    // OPTIONS.poToken = GetPOToken()
    // OPTIONS.visitorData = GetVisitorData()
    // if (!IN_DEVELOPMENT) {
    //     const cookies = GetCookies()
    //     const agent = ytdl.createAgent(cookies)
    //     OPTIONS.agent = agent
    // } else {
    //     OPTIONS.agent = undefined
    // }
    const info = await ytdl.getBasicInfo(url)
    console.log('info')
    if (!info) return false;
    const details = info?.videoDetails
    const seconds = details?.lengthSeconds
    if (!seconds) return false;
    if (+seconds > 900) {
        blockedUrls[url] = 'Video is longer than 15 minutes'
        return 'duration'
    }
    const data = ytdl.download(url, {
        filteringClients: ["web"],
    })
    saveCache(uri, {
        videoId: soundId,
        created: new Date().getTime(),
    })
    return data
}
```

**Reply by 510ren** [reply]
Remove filteringClients from the download function.

**Reply by MOXHARTZ** [reply]
I removed it but it's still the same no difference

**Reply by 510ren** [reply]
Do you still have that problem with the latest version?

**Reply by MOXHARTZ** [reply]
Unfortunately the problem still persists. If you have the motivation to solve the problem, I will not hesitate to give the vds

![image](https://github.com/user-attachments/assets/95b00a49-18d9-4004-bcba-a4805b438134)

**Reply by MOXHARTZ** [reply]
Btw, distube's own ytdl-core was giving the same error. So probably youtube's banning the ip

**Reply by 510ren** [reply]
But if the IP is banned, I think the request to the API will also fail...
Well, that is most likely the case, since the log shows that the request failed.

**Reply by MOXHARTZ** [reply]
But what's even stranger is that I can get song informations from youtube. So I can get the information of the song but it won't let me download it from google's service

**Reply by 510ren** [reply]
Can you tell me the current download code?
I am wondering why the request succeeds but the download fails.

**Reply by 510ren** [reply]
Also, I will create a new issue because this issue needs detailed discussion.

**Reply by 510ren** [reply]
https://github.com/ybd-project/ytdl-core/issues/21

**Reply by 510ren** [reply]
This discussion has been moved to issue and will be resolved.

**Comment by jeh86122** [comment]
What is the difference between web and webcreator in client type?

**Reply by 510ren** [reply]
The WEB client is the usual client used from https://www.youtube.com, etc.
I am not sure about WEB_CREATOR.

**Comment by greyish-coder** [comment]
Hi! Thank you for continuing the ytdl-core project despite all the recent challenges. I have a question about accessing certain format properties after getFullInfo returns an info object. The app I'm working on has a function that selects a suitable format resolution based on user's display:

function getOptimalFormat(info, display, preferredFormat, mediaType, noLog) { ... }

In previous versions of ytdl-core, the 'info' object returned by ytdl.getInfo contained a 'formats' array which contained individual 'format' objects, each of which included a 'height' property (which basically gave me the resolution of that format; e.g. 1080, 1440, etc.) and a 'container' (mp4, webm, etc.) property (so that preferredFormat could be preferred.)

But with the latest ytdl-core version, these properties are no longer there, seemingly? Have they moved somewhere else perhaps?

For context, here's the YTDL_VideoFormat type I found in the youtube.d.ts file in an earlier version (5.0.12) of ytdl-core:

export type YTDL_VideoFormat = {
    itag: number;
    url: string;
    mimeType?: string;
    bitrate?: number;
    audioBitrate?: number;
    width?: number;
    height?: number;
    initRange?: {
        start: string;
        end: string;
    };
    indexRange?: {
        start: string;
        end: string;
    };
    lastModified: string;
    contentLength: string;
    quality: 'tiny' | 'small' | 'medium' | 'large' | 'hd720' | 'hd1080' | 'hd1440' | 'hd2160' | 'highres' | string;
    qualityLabel: '144p' | '144p 15fps' | '144p60 HDR' | '240p' | '240p60 HDR' | '270p' | '360p' | '360p60 HDR' | '480p' | '480p60 HDR' | '720p' | '720p60' | '720p60 HDR' | '1080p' | '1080p60' | '1080p60 HDR' | '1440p' | '1440p60' | '1440p60 HDR' | '2160p' | '2160p60' | '2160p60 HDR' | '4320p' | '4320p60';
    projectionType?: 'RECTANGULAR';
    fps?: number;
    averageBitrate?: number;
    audioQuality?: 'AUDIO_QUALITY_LOW' | 'AUDIO_QUALITY_MEDIUM';
    colorInfo?: {
        primaries: string;
        transferCharacteristics: string;
        matrixCoefficients: string;
    };
    highReplication?: boolean;
    approxDurationMs?: string;
    targetDurationSec?: number;
    maxDvrDurationSec?: number;
    audioSampleRate?: string;
    audioChannels?: number;
    container: 'flv' | '3gp' | 'mp4' | 'webm' | 'ts' | null;
    hasVideo: boolean;
    hasAudio: boolean;
    codecs: string | null;
    videoCodec?: string | null;
    audioCodec?: string | null;
    isLive: boolean;
    isHLS: boolean;
    isDashMPD: boolean;
};

But in the latest version, these properties are no longer there? I guess I'm not sure where to find them or how to access them? 

Please advise!

**Reply by 510ren** [reply]
Thank you. Some of these properties are not included by default because they are not used very often.
To include them, set the option "includesOriginalFormatData" to true.

**Comment by gathilas** [comment]
Am I missing something in the info? Is there a way to get the publish date of the video?

**Reply by 510ren** [reply]
Thank you for your report. It is a bug and we will fix it.

**Reply by 510ren** [reply]
We have released v5.1.8 which fixes this bug and others.

**Comment by iosware** [comment]
What this warning means and how to solve it: "At least one client must be specified."?

**Reply by 510ren** [reply]
This warning is printed if the client options used to retrieve the information are `undefined` or an empty array.
Basically, the solution to this is to either eliminate the client option or specify a non-empty array.
Example:
```ts
// Problem code
const ytdl = new YtdlCore({
    clients: [],
});

// Solution
const ytdl = new YtdlCore({});
/* or */
const ytdl = new YtdlCore({
    clients: ['web', 'mweb'],
});
```

**Comment by ticktricktrack** [comment]
I could use help on what to try next. I've created a simple bot that streams youtube to discord. 

~~edit~~ I was using a data center IP, I now understand that's a fight I'm not going to win

is the proxy example still working? Seems like whatever comes out of the proxy doesn't match any more. 

```
DiscordAPIError[10062]: Unknown interaction
```
